### PR TITLE
Make nested validation rule

### PR DESCRIPTION
**Pull request by xepozz**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | #23

Example of using `Nested` validator:
```php
$values = [
    'author' => [
        'name' => 'Dmitry',
        'age' => 18,
    ],
];
$rules = [
    'author.name' => [
        (new HasLength())->min(3),
    ],
];

$result = (new Nested($rules))->validate($values);
```

Later I'll make upgrade this rule to support deep-nested structure of rule like the following:
```php
$rules = [
  'author' => [
    'name' => [...rules...],
    'age' => [...rules...],
  ],
]
```

### CODE REVIEW

**Comment by xepozz** [comment]
Waiting for https://github.com/yiisoft/arrays/pull/80

File: tests/Rule/NestedTest.php [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            ->errorWhenPropertyPathIsNotFound(true);
+
+        $result = $validator->validate([]);
+
+        $this->assertEquals(['Property path "value" is not found.'], $result->getErrors());
+    }
+
+    public function testPropertyPathIsNotFoundMessage(): void
+    {
+        $validator = (new Nested(['value' => (new Required())]))
```
Comment (roxblnfk) [inline tests/Rule/NestedTest.php:83]:
```suggestion
        $validator = (new Nested(['value' => new Required()]))
```

File: src/Rule/Nested.php [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    private iterable $rules;
+
+    private bool $errorWhenPropertyPathIsNotFound = false;
+    private string $propertyPathIsNotFoundMessage = 'Property path "{path}" is not found.';
+
+    private string $message = 'Nested validation error.';
+
+    public function __construct(iterable $rules)
+    {
+        $rules = $rules instanceof Traversable ? iterator_to_array($rules) : (array)$rules;
```
Comment (roxblnfk) [inline src/Rule/Nested.php:30]:
```suggestion
        $rules = $rules instanceof Traversable ? iterator_to_array($rules) : $rules;
```

File: src/Rule/Nested.php [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    public function getOptions(): array
+    {
+        return $this->rules->asArray();
+    }
+
+    private function checkRules(array $rules): bool
+    {
+        return array_reduce(
+            $rules,
+            fn (bool $carry, $rule) => $carry || is_array($rule) ? $this->checkRules($rule) : !$rule instanceof Rule,
```
Comment (roxblnfk) [inline src/Rule/Nested.php:104]: Are you sure about operators priority?

Comment (xepozz) [reply src/Rule/Nested.php:104]: Sure

File: src/Rule/Nested.php [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    public function propertyPathIsNotFoundMessage(string $message): self
+    {
+        $new = clone $this;
+        $new->propertyPathIsNotFoundMessage = $message;
+        return $new;
+    }
+
+    public function getOptions(): array
+    {
+        return $this->rules->asArray();
```
Comment (samdark) [inline src/Rule/Nested.php:133]: There's no `asArray()` in `rules`. Also, there's no test for this method.

**Review by samdark — CHANGES_REQUESTED** [review]

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
:+1

**Pull request merged**